Add documentation for the auto model updates (#25947)
apache/beam: website/www/site/content/en/documentation/sdks/python-machine-learning.md

For detailed instructions explaining how to build and run a Python pipeline that uses ML models, see the
[Example RunInference API pipelines](https://github.com/apache/beam/tree/master/sdks/python/apache_beam/examples/inference) on GitHub.

> ## Slowly updating side input pattern to update models used by ModelHandler

Reviewer: Is slowly updating a type of side input pattern? If so, I think it should be hyphenated:

Slowly-updating side input pattern

Author: Changed it.

> The RunInference PTransform will accept a side input of [ModelMetadata](https://beam.apache.org/releases/pydoc/current/apache_beam.ml.inference.base.html#apache_beam.ml.inference.base.ModelMetadata), which is a `NamedTuple` containing the `model_id` and `model_name`,

Reviewer: ```suggestion
To update models used by the model handler in the RunInference PTransform without stopping the pipeline, use a [`ModelMetadata`](https://beam.apache.org/releases/pydoc/current/apache_beam.ml.inference.base.html#apache_beam.ml.inference.base.ModelMetadata) side input, which is a `NamedTuple` containing the `model_id` and `model_name`.
```

to update the models used by the ModelHandler in the RunInference PTransform without the need of stopping the pipeline for the model updates.

>   * `model_id`: Unique identifier for the model. This can be a file path or a URL where the model can be accessed. It is used to load the model for inference.

Reviewer 2: `It is used to load the model for inference.` - could you make it explicit that this should be in the same format that the given framework expects for a saved model file (e.g. state_dict for PyTorch)?

  * `model_name`: Human-readable name for the model. This can be used to identify the model in the metrics generated by the RunInference transform.

> **Note**: The side input PCollection must follow [AsSingleton](https://beam.apache.org/releases/pydoc/current/apache_beam.pvalue.html?highlight=assingleton#apache_beam.pvalue.AsSingleton) view or the pipeline will result in error.

Reviewer 2: This is probably hard for the average developer to understand - could we lead with the WatchTransform pattern (ideally with a small code snippet) and then include the generalized ModelMetadata approach as something like "if just using a file watch transform doesn't work for you"

Reviewer 2: Also, maybe you could link to the Watch Transform that you created itself as a pattern to follow?

> **Note**: If the main PCollection emits inputs and side input has yet to receive inputs, the main PCollection will get buffered until there is

Reviewer: We generally don't want two notes in a row. Does this content need to be in a note? Can it just be text in the file? It both sections must be notes, could they be combined into one note?

> **Note**: If the main PCollection emits inputs and side input has yet to receive inputs, the main PCollection will get buffered until there is

Reviewer: ```suggestion
**Note**: If the main PCollection emits inputs and a side input has yet to receive inputs, the main PCollection is buffered until there is
```

>             an update to the side input. This could happen with Global windowed side inputs with data driven triggers such as `AfterCount`, `AfterProcessingTime`. So until there is an update to the side input, emit the default/initial model id that is used to pass the respective `ModelHandler` as side input..

Reviewer 2: ```suggestion
            an update to the side input. This could happen with Global windowed side inputs with data driven triggers such as `AfterCount`, `AfterProcessingTime`. So until there is an update to the side input, emit the default/initial model id that is used to pass the respective `ModelHandler` as side input.
```

nit

>             an update to the side input. This could happen with Global windowed side inputs with data driven triggers such as `AfterCount`, `AfterProcessingTime`. So until there is an update to the side input, emit the default/initial model id that is used to pass the respective `ModelHandler` as side input..

Reviewer: ```suggestion
            an update to the side input. This could happen with Global windowed side inputs with data driven triggers such as `AfterCount`, `AfterProcessingTime`. Until the side input is updated, emit the default or initial model ID that is used to pass the respective `ModelHandler` as a side input.
```

For detailed instructions explaining how to build and run a Python pipeline that uses ML models, see the
[Example RunInference API pipelines](https://github.com/apache/beam/tree/master/sdks/python/apache_beam/examples/inference) on GitHub.

> ## Slowly-updating side input pattern to auto update models used in RunInference

Reviewer: ```suggestion
## Slowly-updating side input pattern to auto-update models used in RunInference
```

> To perform auto updates of the models used in `RunInference` PTransform without stopping the Beam pipeline, pass a [`ModelMetadata`](https://beam.apache.org/releases/pydoc/current/apache_beam.ml.inference.base.html#apache_beam.ml.inference.base.ModelMetadata) side input PCollection to the input parameter `model_metadata_pcoll` of RunInference.

Reviewer: ```suggestion
To perform automatic updates of the models used with the RunInference `PTransform` without stopping the Beam pipeline, pass a [`ModelMetadata`](https://beam.apache.org/releases/pydoc/current/apache_beam.ml.inference.base.html#apache_beam.ml.inference.base.ModelMetadata) side input `PCollection` to the RunInference input parameter `model_metadata_pcoll`.
```

`ModelMetdata` is a `NamedTuple` containing:

>   * `model_id`: Unique identifier for the model. This can be a file path or a URL where the model can be accessed. It is used to load the model for inference. The URL or file path must be in the compatible format so that the respective ModelHandlers can load the models without errors.

Reviewer: ```suggestion
  * `model_id`: Unique identifier for the model. This can be a file path or a URL where the model can be accessed. It is used to load the model for inference. The URL or file path must be in the compatible format so that the respective `ModelHandlers` can load the models without errors.
```

>     **For example**, `PyTorchModelHandler` initially loads a model using weights and a model class. If you pass in weights from a different model class when you update the model using side inputs, the model doesn't load properly, because it expects the weights from the original model class.

Reviewer: ```suggestion
    For example, `PyTorchModelHandler` initially loads a model using weights and a model class. If you pass in weights from a different model class when you update the model using side inputs, the model doesn't load properly, because it expects the weights from the original model class.
```

>     **For example**, `PyTorchModelHandler` loads a model using weights and a model class initially, and during the update, If you pass in weights from a different model class when you update the model via side inputs, it won’t load properly because it’s expecting the weights from the original model class.

Reviewer: ```suggestion
    **For example**, `PyTorchModelHandler` initially loads a model using weights and a model class. If you pass in weights from a different model class when you update the model using side inputs, the model doesn't load properly, because it expects the weights from the original model class.
```

>   * `model_name`: Human-readable name for the model. This can be used to identify the model in the metrics generated by the RunInference transform.

Reviewer: ```suggestion
  * `model_name`: Human-readable name for the model. You can use this name to identify the model in the metrics generated by the RunInference transform.
```

Use cases:

>  * Use `WatchFilePattern` as side input to the RunInference PTransform for the auto updates of the ML model. More details can be found [here](/website/www/site/content/en/documentation/ml/side-input-updates.md).

Reviewer: ```suggestion
 * Use `WatchFilePattern` as side input to the RunInference `PTransform` to automatically update the ML model. For more information, see [Use `WatchFilePattern` as side input to auto-update ML models in RunInference](https://beam.apache.org/documentation/ml/side-input-updates).
```

> The side input PCollection must follow [AsSingleton](https://beam.apache.org/releases/pydoc/current/apache_beam.pvalue.html?highlight=assingleton#apache_beam.pvalue.AsSingleton) view or the pipeline will result in error.

Reviewer: ```suggestion
The side input `PCollection` must follow the [`AsSingleton`](https://beam.apache.org/releases/pydoc/current/apache_beam.pvalue.html?highlight=assingleton#apache_beam.pvalue.AsSingleton) view to avoid errors.
```

> **Note**: If the main PCollection emits inputs and a side input has yet to receive inputs, the main PCollection is buffered until there is

Reviewer: ```suggestion
**Note**: If the main `PCollection` emits inputs and a side input has yet to receive inputs, the main `PCollection` is buffered until there is
```

>             an update to the side input. This could happen with Global windowed side inputs with data driven triggers such as `AfterCount`, `AfterProcessingTime`. Until the side input is updated, emit the default or initial model ID that is used to pass the respective `ModelHandler` as a side input.

Reviewer: ```suggestion
            an update to the side input. This could happen with global windowed side inputs with data driven triggers, such as `AfterCount`, `AfterProcessingTime`. Until the side input is updated, emit the default or initial model ID that is used to pass the respective `ModelHandler` as a side input.
```

>             an update to the side input. This could happen with global windowed side inputs with data driven triggers, such as `AfterCount`, `AfterProcessingTime`. Until the side input is updated, emit the default or initial model ID that is used to pass the respective `ModelHandler` as a side input.

Reviewer: ```suggestion
            an update to the side input. This could happen with global windowed side inputs with data driven triggers, such as `AfterCount` and `AfterProcessingTime`. Until the side input is updated, emit the default or initial model ID that is used to pass the respective `ModelHandler` as a side input.
```
